pkg/utils/kubernetes/status.go

```diff
@@ -0,0 +1,329 @@
+			}
+
+			objLog := log.WithValues(
+				"kind", obj.GetKind(),
+				"resource", ObjectKey(obj.GetName(), obj.GetNamespace()))
+
+			objLog.V(1).Info("get resource status")
+			status, err := Status(obj)
+			if err != nil {
+				allErrs = append(allErrs, err)
+			}
+			objLog.V(1).Info("resource status", "status", status)
+
+			objStates[i] = &ObjectState{
+				Object: obj,
+				Status: status,
+			}
+
+		}(i, obj)
+	}
+	wg.Wait()
+
+	if len(allErrs) > 0 {
+		return nil, apimacherrors.NewAggregate(allErrs)
+	}
+
+	return objStates, nil
+}
+
+// FilterObjectStateInStatusType returns the objects in the desired status.
+func FilterObjectStateInStatusType(objStates ObjectStates, status StatusType) []*unstructured.Unstructured {
+	var objectsInStatus = []*unstructured.Unstructured{}
+	for _, o := range objStates {
+		if o.Status == status {
+			objectsInStatus = append(objectsInStatus, o.Object)
+		}
+	}
+	return objectsInStatus
+}
+
+// FilterObjectStateNotInStatusType returns the objects not in the desired status.
+func FilterObjectStateNotInStatusType(objStates ObjectStates, status StatusType) []*unstructured.Unstructured {
+	var objectsNotInStatus = []*unstructured.Unstructured{}
+	for _, o := range objStates {
+		if o.Status != status {
+			objectsNotInStatus = append(objectsNotInStatus, o.Object)
+		}
+	}
+	return objectsNotInStatus
+}
+
+// Status returns the status for a given object and SatusReady
+// by default for non-managed objects.
+func Status(u *unstructured.Unstructured) (StatusType, error) {
+	gk := u.GroupVersionKind().GroupKind()
+	switch gk.String() {
+	case "Pod":
+		pod := &corev1.Pod{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, pod); err != nil {
+			return StatusUnknown, err
+		}
+		return PodStatus(pod)
+	case "Deployment.apps":
+		dp := &appsv1.Deployment{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, dp); err != nil {
+			return StatusUnknown, err
+		}
+		return DeploymentStatus(dp)
+	case "ReplicaSet.apps":
+		rs := &appsv1.ReplicaSet{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, rs); err != nil {
+			return StatusUnknown, err
+		}
+		return ReplicasetStatus(rs)
+	case "StatefulSet.apps":
+		sts := &appsv1.StatefulSet{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, sts); err != nil {
+			return StatusUnknown, err
+		}
+		return StatefulSetStatus(sts)
+	case "DaemonSet.apps":
+		ds := &appsv1.DaemonSet{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, ds); err != nil {
+			return StatusUnknown, err
+		}
+		return DaemonSetStatus(ds)
+	case "ReplicationController":
+		rc := &corev1.ReplicationController{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, rc); err != nil {
+			return StatusUnknown, err
+		}
+
+		return ReplicationControllerStatus(rc)
+	default:
+		return StatusReady, nil
```

Comment:
The generic why for describing readiness/healthyness of resources is out-of-scope for this PR?

Reply:
I'm not sure to get it, do you mean adding comments to the constants to define what do we consider `Ready`, `InProgress` or `Unknown` ?